### ISSUE TITLE
GPU memory overload render command

### ISSUE THREAD

**Issue by olkovi** [mitsuba-renderer/mitsuba3#483]
System information:

  OS: Windows 10
  CPU: Ryzen 3950x
  GPU: Nvidia 3080Ti
  Python version: 3.10.6
  LLVM version: Not installed
  CUDA version: 11.7
  NVidia driver: 516.94

  Dr.Jit version: ...
  Mitsuba version: 3.1.1
     Compiled with: Wheel from Github
     Variants compiled: Standard

## Description

When using cuda_ad_rgb variant with image = mi.render(scene, spp=256, sensor=sensors[5]), the GPU memory is occupied irreversibly, the next render crashes the kernel

**Comment by njroussel** [comment]
Hi @olkovi 

This is a bit conncerning, could you please provide us with a reproducer? As long as `image` is not deleted, there will of course be some memory utilization, but nothing that should fill up your VRAM. 
Does this also happen in `llvm` variants?

You could also try upgrading your mitsuba/drjit verisons.

**Comment by olkovi** [comment]
Hello, 

I tried to use combine the code from the "Rendering from multiple viewpoints" tutorial with a custom transparent object and scene. Since even in the compressed format, the combined size of the models and textures is 40MB, I will share a link to the full archive: https://drive.google.com/file/d/1HM4gaqnEf7Gm8jVP6PuowNBue-mCfYcn/view?usp=share_link. 
Otherwise the code is as follows:

import mitsuba as mi
import drjit as dr
import torch



device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
print(device)

mi.variants()


mi.set_variant("cuda_ad_rgb")
print(mi.__version__)
import matplotlib.pyplot as plt
from mitsuba import ScalarTransform4f as T



scene = mi.load_file("./scenes/dragontry.xml")





def load_sensor(r, phi, theta):
    return mi.load_dict({
        'type': 'perspective',
        'fov': 39.597755,
        'to_world': T.look_at(
            origin=origin,
            target=[0, 0, 0],
            up=[0, 0, 1]
        ),
        'sampler': {
            'type': 'independent',
            'sample_count': 2048
        },
        'film': {
            'type': 'hdrfilm',
            'width': 2560,
            'height': 1440,
            'rfilter': {
                'type': 'tent',
            },
            'pixel_format': 'rgb',
        },
    })

sensor_count = 8

radius = 3
phis = [20.0 * i for i in range(sensor_count)]
print(phis)
theta = 60

sensors = [load_sensor(radius, phi, theta) for phi in phis]

from matplotlib import pyplot as plt
fig = plt.figure(figsize=(10, 7))
fig.subplots_adjust(wspace=0, hspace=0)
for i in range(sensor_count):
    image = mi.render(scene, spp=4096, sensor=sensors[i])
    # ax = fig.add_subplot(2, 3, i + 1).imshow(image ** (1.0 / 2.2))
    mi.util.write_bitmap("dragon"+str([i])+".png", image)
    plt.axis("off")
    del image

**Comment by njroussel** [comment]
Can you try upgrading your Mitsuba/Dr.Jit verisons?

Also, `del image` might not actually take place right away. You'll want to call `gc.collect()` twice right after that line to guarantee that Python did indeed garbage collect it.

I'm still surprised that this is an issue. If you use smaller resolutions, can you see the memory usage increase for each sensor ?

**Comment by olkovi** [comment]
Hello, 
I attempted the gc.collect() workaround, unfortunately to no avail. By the way, pytorch has a method for clearing VRAM with `torch.cuda.empty_cache()`, might there be something similar in mitsuba, perchance?

It appears that for every new render viewpoint a set amount of memory gets filled in, even though the `image` variable should get overwritten at every iteration. So smaller resolutions (or spp counts) will still progressively occupy the GPU memory, just at a slower rate, until the GPU total is reached and the kernel must be restarted for further renders. 

I will see if the newer version works better than PyPi precompiled 3.1.1. Failing that, I assume running this on a Ubuntu could be tried, and if that works, at least the bug can be localised to Windows.

**Comment by njroussel** [comment]
@olkovi any updates? did the new version fix the issue?
Can I close this issue?

**Comment by olkovi** [comment]
Yes, I've checked. The new 3.2.0 version for Windows did not fix the problem, however it appears to be localised to multi-pass rendering specifically, whether defined automatically or by hand. When multi-pass rendering was unnecessary, the memory could be purged successfully, but when it gets enabled, even small renders eventually fill the memory up. 

Wasn't able to check on Ubuntu just yet, since the Optix apparently didn't match the video drivers.

**Comment by olkovi** [comment]
Unfortunately this means that for high sample count renders the kernel needs to get reset between each render.

**Comment by njroussel** [comment]
I don't know how I missed this until now....

The issue here is that the sensor has some state, namely the sampler. If you call your `load_sensor` method directly in the `render()` function you should no longer have this issue because the sensor is effectively deleted at the end of the loop. Currently, at iteration N of your loop, you'll have N-1 alive sampler states which scale linearly with the `spp`.

**Comment by olkovi** [comment]
This worked! The automatically defined passes per render still killed the kernel, though it appears that the reason was different, as the memory did not get filled, however after adding <integer name="samples_per_pass" value="64"/> under the integrator in the .xml scene parameters, the scene rendered successfully at 12800 samples per pixel.
Thank you!
